fix autocomplete quotes

mikhail-zhadanov opened the pull request:
## Overview

resolves #301

Fixed an issue when autocomplete doesn't consider existing single quotes and provides text with double quotes.
Was: `select * from {{ ref('"model_name"') }}`
Now: `select * from {{ ref('model_name') }}`

As far a I've checked the fix is working for me. As I'm seeing TypeScript for the first time, it would be nice if somebody will point to mistakes.

## Checklist

- [x] I have run this code and it appears to resolve the stated issue
- [ ] `README.md` updated and added information about my change

mdesmet commented:
There is many ways to start/edit the autocompletes, have you tried this in depth?

* starting a new autocomplete, without the jinja code samples `{{ ref(" }}`
* editing an existing autocomplete by removing the inner text `{{ ref("") }}`
* starting to type in an existing autocomplete  `{{ ref("existi__new_text__ng") }}`
* ... ?

Could you let us know which cases you tested?

mikhail-zhadanov commented:
I've tested it again and and it strangely didn't work as during the time when I've submitted this PR.
Now I've done some changes and here's testing result:
![Animation](https://github.com/innoverio/vscode-dbt-power-user/assets/122353008/e957daf8-5d8c-4389-9a87-ee6c198b868c)


During debugging `linePrefix.endsWith` [at line 69](https://github.com/innoverio/vscode-dbt-power-user/blob/89993ae979e9c76fad8970649498d01df9d956ed/src/autocompletion_provider/modelAutocompletionProvider.ts#L61) didn't work as expected. It was not reacting to any correct input provided and  resulted in quoteFound = False in all the cases. Maybe it's related to [String.substr()](https://github.com/innoverio/vscode-dbt-power-user/blob/89993ae979e9c76fad8970649498d01df9d956ed/src/autocompletion_provider/modelAutocompletionProvider.ts#L53C31-L53C31) function being [deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr#browser_compatibility)? Just to be sure I've switched it to String.substring() and now it seems to be working.

Also added some logic to identify quote type if autocomplete was initiated within some point of inner text.

mdesmet commented:
I think you only test the happy path here. What if you are past the ref and you type a quote for example you call another macro with a string argument. We are manipulating the character position by jus searching the first `ref( `in the prefix.

mdesmet commented:
Also in your UI it seems not working as expected. it should be `{{ ref("dbt_artifacts", "dim_dbt__tests") }}

mikhail-zhadanov commented on src/autocompletion_provider/modelAutocompletionProvider.ts:
```diff
@@ -48,16 +48,24 @@ export class ModelAutocompletionProvider
       let quoteFound = false;
       let quote = "";
+      const refString = "ref(";
+      const refStringPosition = linePrefix.indexOf(refString);
```
As far as I understand autocomplete is triggered only if `ref` is present in line as coded here
https://github.com/innoverio/vscode-dbt-power-user/blob/5e43c89c8ad81f519ba2315bf8455eceefbbb446/src/autocompletion_provider/modelAutocompletionProvider.ts#L22

Just quotes without `ref` will not trigger autocomplete. 
In case of `ref` without quotes autocomplete works.

I've compiled extension with current fix and using it daily. So far everything works well.

mikhail-zhadanov commented:
> Also in your UI it seems not working as expected. it should be `{{ ref("dbt_artifacts", "dim_dbt__tests") }}`

This is absolutely correct behaviour of UI. This model compiles successfully. 
<img width="1102" alt="image" src="https://github.com/innoverio/vscode-dbt-power-user/assets/122353008/76748b94-2a06-4ce1-b952-0d6ba2de4937">
 
<img width="982" alt="image" src="https://github.com/innoverio/vscode-dbt-power-user/assets/122353008/f2a77713-b361-45e6-a424-b37f49e2486f">

mdesmet commented:
I think this one is correct {{ ref("dbt_artifacts", "dim_dbt__tests") }}
but this one not {{ ref("dbt_artifacts, dim_dbt__tests") }}

On Sat, Jul 29, 2023 at 3:30 PM Mikhail Zhadanov ***@***.***>
wrote:

> Also in your UI it seems not working as expected. it should be {{
> ref("dbt_artifacts", "dim_dbt__tests") }}
>
> This is absolutely correct behaviour of UI. This model compiles
> successfully.
> [image: image]
> <https://user-images.githubusercontent.com/122353008/256985410-76748b94-2a06-4ce1-b952-0d6ba2de4937.png>
> [image: image]
> <https://user-images.githubusercontent.com/122353008/256985478-f2a77713-b361-45e6-a424-b37f49e2486f.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/innoverio/vscode-dbt-power-user/pull/482#issuecomment-1656731574>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGWCX63CWZ6HFMR3UVMNMLXSUF7PANCNFSM6AAAAAA2V5O6YU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

anandgupta42 commented:
Closing the pull request as it has been open for a while and there are still some unresolved items. This is part of our regular housekeeping process.
